File: ljh/java/src/level1/DartGame.java
package level1;

public class DartGame {
	public int solution(String dartResult) {
		int count = 0;
		int[] arr = new int[3];
		boolean isNumber = false;
		int length = dartResult.length();

		for (int i = 0; i < length; i++) {
			Character kb = dartResult.charAt(i);

			if (kb >= 48 && kb <= 57) {
				// 숫자일때
				arr[count] = Character.getNumericValue(kb);
				if (isNumber == true) {
					arr[count] = 10;
				}
				isNumber = true;
			} else {
				// 숫자가 아닐때
				if (isNumber == true) {// 보너스 일때
					switch (kb) {
					case 83:
						// S 일때
						break;
					case 68:
						// D 일때
						arr[count] = (int) Math.pow(arr[count], 2);
						break;
					case 84:
						// T 일때
						arr[count] = (int) Math.pow(arr[count], 3);
					}
					count = count + 1;
				} else {// 옵션 일때
					if (kb == 42) {
						// *
						if (count > 1) {
							arr[count - 2] = arr[count - 2] * 2;
						}
						arr[count - 1] = arr[count - 1] * 2;
					} else if (kb == 35) {
						// #
						arr[count - 1] = -arr[count - 1];
					}
				}

				isNumber = false;
			}
		}

		int answer = 0;
		for (int j = 0; j < 3; j++) {
			answer = answer + arr[j];
		}
		return answer;
	}
}

File: ljh/javascript/level1/DartGame.js
export default class DartGame {
    solution = (dartResult) => {
        let answer = 0; // 합계
        let num; // 현재 숫자
        let num2 = 0; // 에러 방지
        let arr = []; // 숫자 저장 배열
        let index = 0; //배열 인덱스
        let isNumber = false;  //이전 문자가 숫자인가? 아닌가?
    
        for(let char of dartResult) {
            if(char >= '0' && char <= '9'){
                // 숫자일 때
                if(isNumber === true){
                    num = 10;
                    num2 = 10;
                    continue;
                }
                    num = char;
                    num2 = char;
                    isNumber = true;
            }else{
                // 문자일 때
                isNumber = false;
                switch(char){
                    case 'T':
                        num2 = num2*num2;
                    case 'D':
                        num = num2*num;
                    case 'S':
                        arr.push(num);
                        index++;
                        break;
                    case '#':
                        arr[index-1] *=-1;
                        break;
                    default:
                        if(index>=2){
                            arr[index-2] *= 2;
                        }
                        arr[index-1] *= 2;
                        break;
                }
            }   
        }
        for(let i of arr){
            answer = answer + (+i);
        }
        return answer;
    }
}
